crates/flipperzero/src/furi.rs

```diff
@@ -38,3 +40,60 @@ pub fn sleep(duration: core::time::Duration) {
         }
     }
 }
+
+/// MessageQueue provides a safe wrapper around the furi message queue primitive.
+pub struct MessageQueue<M: Sized> {
+    hnd: *const sys::message_queue::FuriMessageQueue,
+    _marker: core::marker::PhantomData<M>,
+}
+
+impl<M: Sized> MessageQueue<M> {
+    /// Constructs a message queue with the given capacity.
+    pub fn new(capacity: u32) -> Self {
+        Self {
+            hnd: unsafe { sys::message_queue::alloc(capacity, core::mem::size_of::<M>()) },
+            _marker: core::marker::PhantomData::<M>,
+        }
+    }
+
+    // Attempts to add the message to the end of the queue, waiting up to timeout ticks.
+    // FuriStatusOk (0) is returned if the insert is successful, otherwise FuriStatusErrorTimeout
+    // (-2) is returned.
+    pub fn put(&self, mut msg: M, timeout: u32) -> u32 {
+        unsafe {
+            sys::message_queue::put(self.hnd, &mut msg as *mut _ as *const c_void, timeout)
+        }
+    }
+
+    // Attempts to read a message from the front of the queue within timeout ticks.
+    // FuriStatusOk (0) is returned if the read is successful, otherwise FuriStatusErrorTimeout
+    // (-2) is returned.
+    pub fn get(&self, msg: &mut M, timeout: u32) -> u32 {
```

Comment:
Similarly, this would return something like `Result<M, furi::Error>`, rather than using a `&mut M` parameter.

Reply:
Will do!